src/measure_theory/function/jacobian.lean

```diff
@@ -0,0 +1,1274 @@
+  (h'f : measurable f) :
+  measure.map f ((μ.restrict s).with_density (λ x, ennreal.of_real (|(f' x).det|)))
+    = μ.restrict (f '' s) :=
+begin
+  apply measure.ext (λ t ht, _),
+  rw [map_apply h'f ht, with_density_apply _ (h'f ht), measure.restrict_apply ht,
+      restrict_restrict (h'f ht),
+      lintegral_abs_det_fderiv_eq_add_haar_image μ ((h'f ht).inter hs)
+        (λ x hx, (hf' x hx.2).mono (inter_subset_right _ _)) (hf.mono (inter_subset_right _ _)),
+      image_preimage_inter]
+end
+
+/-- Change of variable formula for differentiable functions, set version: if a function `f` is
+injective and differentiable on a measurable set `s`, then the pushforward of the measure with
+density `|(f' x).det|` on `s` is the Lebesgue measure on the image set. This version is expressed
+in terms of the restricted function `s.restrict f`.
+For a version for the original function, but with a measurability assumption,
+see `map_with_density_abs_det_fderiv_eq_add_haar`.
+-/
+theorem restrict_map_with_density_abs_det_fderiv_eq_add_haar
+  {f : E → E} {s : set E} (hs : measurable_set s)
+  {f' : E → (E →L[ℝ] E)} (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hf : inj_on f s) :
+  measure.map (s.restrict f)
+    (comap coe (μ.with_density (λ x, ennreal.of_real (|(f' x).det|)))) = μ.restrict (f '' s) :=
+begin
+  obtain ⟨u, u_meas, uf⟩ : ∃ u, measurable u ∧ eq_on u f s,
+  { classical,
+    refine ⟨piecewise s f 0, _, piecewise_eq_on _ _ _⟩,
+    refine continuous_on.measurable_piecewise _ continuous_zero.continuous_on hs,
+    have : differentiable_on ℝ f s := λ x hx, (hf' x hx).differentiable_within_at,
+    exact this.continuous_on },
+  have u' : ∀ x ∈ s, has_fderiv_within_at u (f' x) s x :=
+    λ x hx, (hf' x hx).congr (λ y hy, uf hy) (uf hx),
+  set F : s → E := u ∘ coe with hF,
+  have A : measure.map F
+    (comap coe (μ.with_density (λ x, ennreal.of_real (|(f' x).det|)))) = μ.restrict (u '' s),
+  { rw [hF, ← measure.map_map u_meas measurable_subtype_coe, map_comap_subtype_coe hs,
+        restrict_with_density hs],
+    exact map_with_density_abs_det_fderiv_eq_add_haar μ hs u' (hf.congr uf.symm) u_meas },
+  rw uf.image_eq at A,
+  have : F = s.restrict f,
+  { ext x,
+    exact uf x.2 },
+  rwa this at A,
+end
+
+/-! ### Change of variable formulas in integrals -/
+
+/- Change of variable formula for differentiable functions: if a function `f` is
+injective and differentiable on a measurable set `s`, then the Lebesgue integral of a function
+`g : E → ℝ≥0∞` on `f '' s` coincides with the integral of `|(f' x).det| * g ∘ f` on `s`.
+Note that the measurability of `f '' s` is given by `measurable_image_of_fderiv_within`. -/
+theorem lintegral_image_eq_lintegral_abs_det_fderiv_mul {f : E → E} {s : set E}
+  (hs : measurable_set s) {f' : E → (E →L[ℝ] E)}
+  (hf' : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hf : inj_on f s) (g : E → ℝ≥0∞) :
+  ∫⁻ x in f '' s, g x ∂μ = ∫⁻ x in s, ennreal.of_real (|(f' x).det|) * g (f x) ∂μ :=
```

Comment:
Could also consider stating this as
```suggestion
  ∫⁻ x in f '' s, g x ∂μ = ∫⁻ x in s, ∥(f' x).det∥₊ * g (f x) ∂μ :=
```
but maybe that's more annoying to work with

Reply:
Yes, that's more annoying. The interest of `ennreal.of_real` instead of the composition of the two maps from real to nnreal and then from nnreal to ennreal is that it has a good API.